ui/components/app/detected-token/detected-token.js

```diff
@@ -0,0 +1,123 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import { useSelector, useDispatch } from 'react-redux';
+import { chain } from 'lodash';
+
+import {
+  importTokens,
+  ignoreTokens,
+  setNewTokensImported,
+} from '../../../store/actions';
+import { getDetectedTokensInCurrentNetwork } from '../../../selectors';
+
+import DetectedTokenSelectionPopover from './detected-token-selection-popover/detected-token-selection-popover';
+import DetectedTokenIgnoredPopover from './detected-token-ignored-popover/detected-token-ignored-popover';
+
+const DetectedToken = ({ setShowDetectedTokens }) => {
+  const dispatch = useDispatch();
+
+  const detectedTokens = useSelector(getDetectedTokensInCurrentNetwork);
+
+  const [tokensListDetected, setTokensListDetected] = useState(() =>
+    detectedTokens.reduce((tokenObj, token) => {
+      tokenObj[token.address] = { token, selected: true };
+      return tokenObj;
+    }, {}),
+  );
+  const [
+    showDetectedTokenIgnoredPopover,
+    setShowDetectedTokenIgnoredPopover,
+  ] = useState(false);
+
+  const handleClearTokensSelection = async () => {
+    // create a lodash chain on this object
+    const { selected: selectedTokens, deselected: deSelectedTokens } = chain(
+      tokensListDetected,
+    )
+      // get the values
+      .values()
+      // create a new object with keys 'selected', 'deselected' and group the tokens
+      .groupBy((token) => (token.selected ? 'selected' : 'deselected'))
+      // ditch the 'selected' property and get just the tokens'
+      .mapValues((group) => group.map(({ token }) => token))
+      // Exit the chain and get the underlying value, an object.
+      .value();
+
+    if (deSelectedTokens.length < detectedTokens.length) {
+      await dispatch(ignoreTokens(deSelectedTokens));
+      await dispatch(importTokens(selectedTokens));
+      const tokenSymbols = selectedTokens.map(({ symbol }) => symbol);
+      dispatch(setNewTokensImported(tokenSymbols.join(', ')));
+    } else {
+      await dispatch(ignoreTokens(deSelectedTokens));
+    }
+    setShowDetectedTokens(false);
+  };
+
+  const handleTokenSelection = (token) => {
+    setTokensListDetected((prevState) => ({
+      ...prevState,
+      [token.address]: {
+        ...prevState[token.address],
+        selected: !prevState[token.address].selected,
+      },
+    }));
+  };
+
+  const onImport = async () => {
+    // create a lodash chain on this object
+    const { selected: selectedTokens } = chain(tokensListDetected)
```

Comment:
Is this the same chain routine as the previous?  Maybe we create a helper function for this?